### PR TITLE
Feature: Retry deploying functions that failed to deploy

### PR DESCRIPTION
**Pull request by jpreynat**
### Description

This PR adds a new `--retry <times>` flag to the `firebase deploy` command.
It automatically retries to deploy functions that failed to deploy, in the case that:
- they didn't fail to deploy because of a quota error
- all functions didn't fail to deploy

The aim is to reduce friction when deploying many functions, especially in a CI environment, and reduce deployment time, since the source code doesn't need to be parsed again and re-uploaded.
	 
### Scenarios Tested

I've tested the command successfully in the basic cases.
I've also tested it against a quota error to ensure that it doesn't try to redeploy the functions in this case.

It also correctly handles namespacing, ie:

If the project includes 3 functions `ns.fn1`, `ns.fn2` and `main`, runnning:
```
firebase deploy --retry 3 --only functions:ns
```
will only deploy `ns.fn1` and `ns.fn2`.
If only one of them fails to deploy, it will only try to re-deploy the one that failed.

### Sample Commands

`firebase deploy --retry 3 --only functions:main`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* Ensures `auth:export` results are fully flushed to the output file.
```
Comment (bkendall) [inline CHANGELOG.md]: Why is this bundled in with this PR?

File: package-lock.json [hunk 2]
```diff
@@ -157,6 +157,43 @@
       "integrity": "sha512-AbK4ZPP0MMXMuUzfxQb3f71kA++kV9wao0ps5jOsvaHXk1SjS+0Cxsfae+wqfwyvy1oyF9N6Kl8zYqtJizQmjA==",
       "dev": true
     },
+    "@firebase/app": {
```
Comment (bkendall) [inline package-lock.json]: I don't see a reason this file should have been updated... can this be reverted?

File: CHANGELOG.md (hunk 1 above)
Comment (jpreynat) [reply CHANGELOG.md]: Sorry, this must be a leftover from an older PR I made.
I'm gonna clean this.

**Comment by jpreynat** [comment]
@bkendall I rollbacked unwanted changes and updated the branch from the latest master.

**Comment by jpreynat** [comment]
@bkendall any other blockers to merge this PR?

**Comment by razorRun** [comment]
Any Update on this guys? I used to get this error once in a while. But since the last few days, I have been getting this on every deploy from my CICD pipeline.

**Comment by mbleigh** [comment]
Hey folks, please follow along here: https://issuetracker.google.com/issues/154260223

We're aware of reports of increased deploy failures in the last couple weeks and the Cloud Functions team is investigating. We can separately investigate retry logic in the CLI, but the backend is the focus at the moment. Thanks!

**Comment by jpreynat** [comment]
@mbleigh As I mentioned in #853, the timeout to build and deploy cloud functions on GCP has recently (2 to 3 weeks) been reduced from 10 to 5 minutes.

This may have impacted lots of Firebase users, since the install time for node modules can easily take more than 5 minutes when running from scratch.
GCP caches node modules between deployment based on the hash of either `package-lock.json` or `yarn.lock` file, so a simple dependency update leads to re-install all of them.

Installing and building native modules can be pretty long, some other modules run postinstall scripts. `puppeteer` for instance needs to download the Chromium binary that is ~140MB...

We fixed this by building our own tooling to deploy functions separately, bundle all non-native dependencies in our main function code using webpack, and trimming their `package.json` files to leave only the non-bundable dependencies to install.
But this is a bit worrying that GCP makes such an update to its build system without alerting its users, and takes so much time to investigate and understand the source of such an issue.

**Comment by alexlouden** [comment]
Really big 👍  for this one - we have >50 functions and our CI deploy fails more often than not (due to errors or quota), a retry would make it a tonne more reliable

**Comment by rhodgkins** [comment]
Has this been forgotten about? We get this error almost consistently when deploying from Cloud Build and its really frustrating as we have to start our whole release process off again.

**Comment by devth** [comment]
I opened https://issuetracker.google.com/issues/160037033 a few weeks ago with a [repro code sample](https://github.com/devth/cloud-functions-build-error). They said they are working on a fix:

> The Cloud Function team is still working on this issue to increase the timeout for underlying builds for Cloud functions. there is no ETA for the implementation at the time being.

> Aug 7, 2020

**Comment by joehan** [comment]
Hey all, I recently did a major refactor of functions deploy, and I think it should address a lot of the issues you've all reported here. Update to the latest version and check it out :)

I'm going to close this PR since the underlying code changed so much that its quite out of date. If you're still having issues after updating to the latest version, please open an issue on this repo.

**Pull request closed without merging**